### DISCUSSION TITLE
Writing png file loses transparency when saving

### DISCUSSION THREAD

**Discussion by birchcode** [ImageMagick/ImageMagick#6103] [Help] [answered]
### ImageMagick version

7.1.0-59

### Operating system, version and so on

Osx 11.6, PHP 8.2, Imagick 3.7

### Description

A little counter intuitive issue with PHP image magick. 

I'm opening a blankcanvas which is just a transparent png file, then saving it and the written file is white.
```
$im = new \Imagick(realpath('blankcanvas.png'));
$write = $im->writeImage ( 'newfile.png' );
```

This should work in my opinion the same as it does using convert in the cli
```
convert blankcanvas.png -resize 256x256 resized.png
```
resized.png is transparent.

I assume there must be some weird meta data in my png which php imagick doesn't pickup?
```
magick identify blankcanvas.png
blankcanvas.png PNG 64x64 64x64+0+0 8-bit sRGB 109B 0.000u 0:00.000
```
Maybe that is supposed to say sRGBA? 

Any ideas on why this simple process gets different results?

**Comment by snibgo** [comment]
Please post your blankcanvas.png.

> $img = new \Imagick(realpath('blankcanvas.png'));
$write = $im->writeImage ( 'newfile.png' );

Why does the first line have `$img` and the second line have `$im`?

**Reply by birchcode** [reply]
Sorry, they aren't the variables I'm using. I simplified my code here for brevity and messed it up.

Please see blankcanvas.png attached
![blankcanvas](https://user-images.githubusercontent.com/1013599/221097869-3393c7a8-58b8-47bc-84e0-54b2ad821ad2.png)


I should also add that setBackgroundColor does nothing.
```
$im->setBackgroundColor(new ImagickPixel('transparent'));
```

**Comment by snibgo** [comment]
I can't see any problem with `blankcanvas.png`. All 64x64 pixels are transparent black.

I don't use IMagick so can't comment on the code, but perhaps you have over-simplified it, and you are actually doing some other operation(s).

**Comment by birchcode** [comment]
You are quite right. I have over-simplified! Thank you for that.

I stripped down my code and realised it is a mergeLayers operation that is losing the transparency.

The following produces a transparent image
```
$im = new \Imagick(realpath($PROJECT_DIR . '/blankcanvas.png'));
$write = $im->writeImage ( $PROJECT_DIR . '/test1.png' );
```

While this gives me a white image
```
$im = new \Imagick(realpath($PROJECT_DIR . '/blankcanvas.png'));
$result = $im->mergeImageLayers(imagick::LAYERMETHOD_UNDEFINED);
$write = $result->writeImage ( $PROJECT_DIR . '/test2.png' );
```

I have tried a few other Layer Method Constants to no avail and I need to use mergeImageLayers since I'm combining a few images which also have transparent backgrounds. They all compile fine retaining the transparency for each layer but the background ends up white.

**Comment by snibgo** [comment, marked as answer]
I suppose IMagick `mergeImageLayers` is the same as CLI `-layers merge`. On that, See https://imagemagick.org/script/command-line-options.php#layers , the "flatten" and "merge" sections. The background colour is relevant, and defaults to white.

**Reply by birchcode** [reply]
Thanks for this.

As a workaround, I ditched the mergeLayers method and did a much slower method with compositeImage (https://www.php.net/manual/en/imagick.compositeimage.php), which just lets you merge 2 images together at a time, so I am writing and saving to the same file repeatedly in a loop until all the layers are merged. Very time consuming but the transparency works.

A shame there is a bit of a discrepancy between the documentation / implementation of the cli and php imagick. Those layermethod constants aren't even really documented in php docs. 

I can see in the CLI docs with the 'flatten' option it removes the transparency. Trying the constant LAYERMETHOD_MERGE doesn't seem to have any effect on my end. Maybe it's just me.

**Comment by snibgo** [comment]
If you leave the background colour as white, or set it to any opaque colour, merging layers will remove transparency. But if you set it to "none" or "transparent" (they mean the same thing), transparency will not be removed.

PHP IMagick is a separate system to ImageMagick, and is maintained by different people.

**Reply by birchcode** [reply]
Thank you.  I extra appreciate the help you've given
